File: architecture-skills/object-oriented-programming-paradigm/examples/open-closed/open-closed.cs
// Open/Closed Principle in C#.
//
// PricingEngine is closed: it iterates registered DiscountRule
// implementations and never names a concrete rule. FlashSale is
// added later as a new class -- zero edits to the engine.

using System;
using System.Collections.Generic;

public sealed record Order(double Subtotal, int ItemCount, bool IsMember);

/// <summary>Extension point: one implementation per business rule.</summary>
public interface IDiscountRule
{
    string Name { get; }
    double Discount(Order order);
}

// --- Extensions: the engine never mentions these types ---

public sealed class BulkDiscount : IDiscountRule
{
    public string Name => "bulk";
    public double Discount(Order o) =>
        o.ItemCount >= 10 ? 0.05 * o.Subtotal : 0.0;
}

public sealed class MemberDiscount : IDiscountRule
{
    public string Name => "member";
    public double Discount(Order o) =>
        o.IsMember ? 0.10 * o.Subtotal : 0.0;
}

/// <summary>Closed core: computes a total over whatever rules it was given.</summary>
public sealed class PricingEngine
{
    private readonly List<IDiscountRule> _rules = new();

    public void Register(IDiscountRule rule) => _rules.Add(rule);

    public double Total(Order order)
    {
        double total = order.Subtotal;
        foreach (var rule in _rules)
        {
            double amount = rule.Discount(order);
            if (amount > 0.0)
                Console.WriteLine($"  {rule.Name,6}: -${amount:F2}");
            total -= amount;
        }
        return total;
    }
}

/// <summary>A NEW rule, written later -- nothing above is touched.</summary>
public sealed class FlashSale : IDiscountRule
{
    private readonly double _amount;
    private readonly double _threshold;

    public FlashSale(double amount, double threshold)
    {
        _amount = amount;
        _threshold = threshold;
    }

    public string Name => "flash";
    public double Discount(Order o) =>
        o.Subtotal > _threshold ? _amount : 0.0;
}

public static class Program
{
    public static void Main()
    {
        var order = new Order(120.00, 12, true);

        var engine = new PricingEngine();
        engine.Register(new BulkDiscount());
        engine.Register(new MemberDiscount());

        Console.WriteLine($"subtotal: ${order.Subtotal:F2}");
        Console.WriteLine($"total:    ${engine.Total(order):F2}");

        Console.WriteLine("-- registering FlashSale: no engine edits --");
        engine.Register(new FlashSale(15.00, 100.00));
        Console.WriteLine($"total:    ${engine.Total(order):F2}");
    }
}

File: architecture-skills/object-oriented-programming-paradigm/examples/open-closed/open-closed.php
<?php
/**
 * Open/Closed Principle in PHP.
 *
 * PricingEngine is closed: it iterates registered discount rules and
 * never names a concrete rule. FlashSale is added later with zero edits
 * to the engine.
 */

declare(strict_types=1);

final readonly class Order
{
    public function __construct(
        public float $subtotal,
        public int $itemCount,
        public bool $isMember,
    ) {}
}

interface DiscountRule
{
    public function name(): string;
    public function discount(Order $order): float;
}

final class BulkDiscount implements DiscountRule
{
    public function name(): string { return 'bulk'; }

    public function discount(Order $order): float
    {
        return $order->itemCount >= 10 ? 0.05 * $order->subtotal : 0.0;
    }
}

final class MemberDiscount implements DiscountRule
{
    public function name(): string { return 'member'; }

    public function discount(Order $order): float
    {
        return $order->isMember ? 0.10 * $order->subtotal : 0.0;
    }
}

final class PricingEngine
{
    /** @var list<DiscountRule> */
    private array $rules = [];

    public function register(DiscountRule $rule): void
    {
        $this->rules[] = $rule;
    }

    public function total(Order $order): float
    {
        $total = $order->subtotal;
        foreach ($this->rules as $rule) {
            $amount = $rule->discount($order);
            if ($amount > 0.0) {
                printf("  %6s: -$%.2f\n", $rule->name(), $amount);
            }
            $total -= $amount;
        }
        return $total;
    }
}

final class FlashSale implements DiscountRule
{
    public function __construct(
        private readonly float $amount,
        private readonly float $threshold,
    ) {}

    public function name(): string { return 'flash'; }

    public function discount(Order $order): float
    {
        return $order->subtotal > $this->threshold ? $this->amount : 0.0;
    }
}

$order = new Order(120.00, 12, true);
$engine = new PricingEngine();
$engine->register(new BulkDiscount());
$engine->register(new MemberDiscount());

printf("subtotal: $%.2f\n", $order->subtotal);
printf("total:    $%.2f\n", $engine->total($order));

echo "-- registering FlashSale: no engine edits --\n";
$engine->register(new FlashSale(15.00, 100.00));
printf("total:    $%.2f\n", $engine->total($order));
